Helper/Data.php

```diff
@@ -25,35 +25,67 @@ public function __construct(
+        $number = ltrim($number, '0');
 
-        $phoneCodes = PhoneCodes::$phoneCodes;
-        $phoneCode = '';
-        if ($phoneCodes) {
-            if (array_key_exists($country, $phoneCodes)) {
-                $phoneCode = $phoneCodes[$country];
-            }
-        }
+        if (!empty($countryCode)) {
+            $prefix = $this->getCountryDigitPrefixFromArray($countryCode);
+            if (!empty($prefix)) {
+                $numberCheck = ltrim($number, '+');
+                $numberCheck = ltrim($numberCheck, '0');
+                $prefixCheck = ltrim($prefix, '+');
 
-        $phoneCodeChk =  (string) preg_replace("/[^0-9]/", "", $phoneCode);
-        $phone =  (string) preg_replace("/[^0-9]/", "", $phone);
-        $prefix = substr($phone, 0, 2);
+                $formattedNumber = null;
 
-        if (strpos($phone, $phoneCodeChk) === 0 || $prefix == '00') {
-            $formattedPhone = '+'.$phone;
-        } else {
-            $formattedPhone = '+'.$phoneCode.$phone;
+                /* Discard invalid phone numbers */
+                if (strlen($numberCheck) < 7 || strlen($numberCheck) > 15) {
+                    return null;
+                }
+                if (is_numeric($number)) {
+                    if ((strpos($number, '+') === 0 || strpos($number, '00') === 0)
+                        && strpos($numberCheck, $prefixCheck) === 0
+                    ) {
+                        $formattedNumber = '+' . $numberCheck;
+                    } elseif ((strpos($number, '+') === 0 || strpos($number, '00') === 0)
+                        && strpos($numberCheck, $prefixCheck) !== 0
+                    ) {
+                        $formattedNumber = '+' . $prefixCheck . $numberCheck;
+                    } elseif (strpos($numberCheck, $prefixCheck) !== 0
+                        || (strtolower($countryCode) == 'no' && strlen($numberCheck) <= 8)
+                    ) {
+                        $formattedNumber = '+' . $prefix . $numberCheck;
+                    } elseif (strpos($numberCheck, $prefixCheck) === 0
+                        && strpos($numberCheck, '0') === strlen($prefixCheck)
+                    ) {
+                        $formattedNumber = ltrim($numberCheck, $prefix);
+                        $formattedNumber = ltrim($formattedNumber, '0');
+                        $formattedNumber = '+' . $prefixCheck . $formattedNumber;
+                    } elseif ((strpos($number, '+') !== 0) && (strpos($number, '00') !== 0)
+                        && strpos($numberCheck, $prefixCheck) === 0
+                    ) {
+                        $formattedNumber = '+' . $number;
+                    } else {
+                        $formattedNumber = null;
+                    }
+                }
+                return $formattedNumber;
+            } else {
+                return $number;
+            }
         }
-
-        return $formattedPhone;
+        return $number;
```

Comment:
when returning `$number`, can it be not in the right length or contain other bad chars?

Reply:
Actually it will be sanitised for special chars in the beginning of the method. 
Probably we can add a length check for the final version of the phone in the end of the method.

Reply:
Magento do not allow `number` to have other chars than `[+-() 0-9]`?

Reply:
I think that also when we return `$formattedNumber = '+' . $prefixCheck . $numberCheck;`
the final number can be longer then 15

as @eugeenek-s said maybe we can just check the final number with regex `+[0-9]{7,15}`

Reply:
No but they will be replaced on lines 34-37

Reply:
@niraizik I agree with your proposal